src/main/java/org/radarcns/redcap/listener/TokenManagerListener.java

```diff
@@ -44,54 +41,38 @@
     private static final Logger LOGGER = LoggerFactory.getLogger(TokenManagerListener.class);
 
     private static final String ACCESS_TOKEN = "TOKEN";
-    private static final String AUTHORIZATION_HEADER = "Authorization";
-    private static final String SCOPE_HEADER = "scope";
-
-    private static final OkHttpClient client;
-    private static final Request request;
 
-    private static long tokenDuration;
-    private static long lastRefresh;
+    private static final OAuth2Client client;
+    private static OAuth2AccessToken token;
 
     static {
-        client = new OkHttpClient.Builder()
-            .connectTimeout(10, TimeUnit.SECONDS)
-            .writeTimeout(10, TimeUnit.SECONDS)
-            .readTimeout(30, TimeUnit.SECONDS)
-            .authenticator((route, response) -> response.request().newBuilder()
-                    .header(AUTHORIZATION_HEADER, Properties.getOauthCredential())
-                    .build()
-            ).build();
-
         try {
-            request = new Request.Builder()
-                .url(Properties.getTokenEndPoint())
-                .post(new FormBody.Builder().add("grant_type", "client_credentials")
-                    //TODO check if scope is correct. This token must be able to access projects
-                    // and create users
-                    .add(SCOPE_HEADER, "read")
-                    .build())
-                .build();
-        } catch (MalformedURLException exc) {
-            LOGGER.error("Properties cannot be load. Check the log for more information.", exc);
+            client = new OAuth2Client()
+                        .managementPortalUrl(Properties.getTokenEndPoint().toString())
+                        .clientId(Properties.getOauthClientId())
+                        .clientSecret(Properties.getOauthClientSecret())
+                        .addScope("read")
+                        .addScope("write");
+        }
+        catch (MalformedURLException exc) {
+            LOGGER.error("Properties cannot be loaded. Check the log for more information.", exc);
             throw new ExceptionInInitializerError(exc);
         }
+        token = new OAuth2AccessToken();
     }
 
     @Override
     public void contextInitialized(ServletContextEvent sce) {
         try {
             getToken(sce.getServletContext());
-        } catch (IOException exc) {
-            LOGGER.warn("{} cannot be generated.", ACCESS_TOKEN);
+        } catch (IllegalStateException exc) {
```

Comment:
Does `getToken` raises only `IllegalStateException`? What happen if `client.getAccessToken()` fails in `getToken`?

Reply:
Indeed, well at first `client.getAccessToken()` would not throw any exceptions. Instead, the `OAuth2AccessToken` could be checked for errors manually with the `isValid()` and `getError()` and `getErrorDescription()` methods. This is because if we can reach MP, but something is wrong with our request, it will respond with a valid JSON describing the error. So we need fields in `OAuth2AccessToken` to capture these optional JSON fields. However, it is indeed better if the cilent of the class does not need to check validity manually, so i changed the implementation so that `OAuth2Client` will throw a `TokenException`, not only on `IO` or `ObjectMapper` exceptions, but also on error responses from MP.

Reply:
Yep, I agree! ;)